### PR TITLE
Challenge two pull request 

### PR DESCRIPTION
**Pull request by mwaz**
# What does this PR do
The pull request enables one to create recipe categories, to update recipe categories and to delete recipe categories. Each of the recipe categories has recipes which can also be deleted and updated. The PR has two classes Recipe and Users, the Recipe class holds methods for objects that inherit the Recipe class capabilities including creating recipe categories, updating the recipe categories and even deleting the recipe categories. The Users class, on the other hand, contains methods that are used to create objects for the users such as the registration of a new member and the member login functionalities 

# Descriprion of task to be completed 
The task to be completed is a Test-driven development approach that uses OOP to create objects which can then be utilized by methods that call the classes

# How should this be manually tested
One should clone the repo, install a virtual environment and install the files required to run the app in requirements.txt

# Any background context you want to provide?
The APP utilizes CRUD, basically, it does Create, Update, Delete and Read data from objects that are stored in data structures either lists, dictionaries or even tuples.

# What are the relevant pivotal tracker stories?
Relevant pivotal tracker stories can be found here -> https://www.pivotaltracker.com/n/projects/2119558

### CODE REVIEW

**Review by MuthomiMate — APPROVED** [review]

**Comment by kaguna** [comment]
Elegant commit messages

File: .gitignore [hunk 1]
```diff
@@ -1 +1,9 @@
-.idea
\ No newline at end of file
+.idea
```
Comment (TinaMurimi) [inline .gitignore:1]: Your .coveragerc should be here

File: .gitignore [hunk 2]
```diff
@@ -1 +1,9 @@
...
+.coverage
+Pipfile
+Pipfile.lock
+__pycache__
+app/__pycache__
```
Comment (TinaMurimi) [inline .gitignore:6]: Can be simplified to */__pycache__

File: README.md [hunk 3]
```diff
@@ -11,21 +14,51 @@ UML diagrams for the Project
...
+
+# Install the application's dependencies from requirements.txt to the virtual environment
+
+$ (yourenvname) pip install -r requirements.txt
+
+# Set up Unit Test Environment
+
+run the following command to install nose unit testing environment:
+
+$ (yourenvname) pip install nose
```
Comment (TinaMurimi) [inline README.md:50]: `(yourenvname)` should not be included. Instead, have steps for installing and setting up the virtual environment.

File: Procfile [hunk 4]
```diff
@@ -0,0 +1,2 @@
+web: python run.py
+heroku ps:scale web=1
```
Comment (TinaMurimi) [inline Procfile:2]: Why is this here?

File: recipe.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
+import re
+regex_name = "[a-zA-Z0-9- .]+$"
```
Comment (TinaMurimi) [inline recipe.py:2]: Why is this outside the class?

File: recipe.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        self.cat_name = cat_name
+        self.owner = owner
+        self.recipe_name = recipe_name
+
+    def category_register(self, cat_name, owner):
+        """ method that defines the elements required to create an account """
+        if re.match(regex_name, cat_name):
+            if cat_name != '' and cat_name.strip():
+                if cat_name not in self.recipe_categories :
+                    self.recipe_categories[cat_name] = {"cat_name":cat_name, "owner":owner,}
```
Comment (TinaMurimi) [inline recipe.py:21]: Set up linters to avoid long lines of code

File: recipe.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+    def __init__(self, cat_name=None, owner=None, recipe_name=None):
+        """ Initializing  class instance variables"""
+        self.cat_name = cat_name
+        self.owner = owner
+        self.recipe_name = recipe_name
+
+    def category_register(self, cat_name, owner):
+        """ method that defines the elements required to create an account """
+        if re.match(regex_name, cat_name):
```
Comment (TinaMurimi) [inline recipe.py:18]: `re.match()` vs `re.search()`. This may cause a problem because `re.match()` matches only from the beginning of the string. https://stackoverflow.com/questions/180986/what-is-the-difference-between-pythons-re-search-and-re-match

File: recipe.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+            if cat_name != '' and cat_name.strip():
```
Comment (TinaMurimi) [inline recipe.py:19]: You are checking if `cat_name` is not an empty string and not on if `cat_name` is `None`. Where is this handled?

File: recipe.py (hunk 6 above)
Comment (TinaMurimi) [reply recipe.py:21]: Use ternary statements

File: recipe.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        """ method that defines the elements required to create an account """
+        if cat_name in self.recipe_categories.keys() :
+            del self.recipe_categories[cat_name]
+            return "200,OK"
+        return "404,Category doesnt exist"
+
+    def recipe_register(self, cat_name, recipe_name, owner):
+        """ method that defines the elements required to create an account """
+        if re.match(regex_name, cat_name):
+            if cat_name != '' and cat_name.strip():
```
Comment (TinaMurimi) [inline recipe.py:37]: What if `cat_name` is None?

File: recipe.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    def category_delete(self, cat_name):
+        """ method that defines the elements required to create an account """
+        if cat_name in self.recipe_categories.keys() :
+            del self.recipe_categories[cat_name]
+            return "200,OK"
+        return "404,Category doesnt exist"
+
+    def recipe_register(self, cat_name, recipe_name, owner):
+        """ method that defines the elements required to create an account """
+        if re.match(regex_name, cat_name):
```
Comment (TinaMurimi) [inline recipe.py:36]: `re.match()` vs `re.search()`?

File: recipe.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+            if cat_name != '' and cat_name.strip():
+                if recipe_name not in self.recipes:
+                    self.recipes[recipe_name] = {"cat_name": cat_name, "recipe_name": recipe_name, "owner": owner}
+                    return "200,OK"
+                return "204,Recipe exists"
+            return "205,Invalid Name"
+        return "205,Regex mismatch"
```
Comment (TinaMurimi) [inline recipe.py:43]: What is this for?

File: recipe.py [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    def recipe_delete(self,recipe_name):
+        """ method that defines the elements required to create an account """
+        if recipe_name in self.recipes.keys():
+            del self.recipes[recipe_name]
+            return "200,OK"
+        return "404,Recipe doesnt exist"
+
+    def view_recipe_category(self,owner):
+        category_data = self.recipe_categories
+        render_category = []
```
Comment (TinaMurimi) [inline recipe.py:54]: Name your variables in such a way another person can understand what that variable stores

File: test_recipes.py [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+    """ Class performing unit testing for class Recipe"""
+    def setUp(self):
+        """Defining setUp() method that runs prior to each test."""
+        self.newCat = Recipe()
+
+    def test_category_registration(self):
+        """ Test to test function create category """
+        self.newCat.recipe_categories = {}
+        self.newCat.category_register("category_one", "waweru@gmail.com")
+        category_registration = self.newCat.category_register("category", "waweru@gmail.com")
```
Comment (TinaMurimi) [inline test_recipes.py:15]: `self.newCat.category_register("category", "waweru@gmail.com")`is used more than once in your code. DRY! That is why the `setUp()` method is used.

File: test_users.py [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+    """
+    def setUp(self):
+        """ Method called to prepare the test fixture """
+        self.newUser = Users()
+
+    def test_member_register(self):
+        """ Test to test function member_register """
+        self.newUser.users = {}
+        #users dictionary to check whether the user already exists
+        self.newUser.member_register("waweru@gmail.com", "mwaz", "password", "password")
```
Comment (TinaMurimi) [inline test_users.py:17]: This should be in the `setUp()` method in case it is being reused

File: test_users.py [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+        """ Test to check if cpasword is null """
+        result = self.newUser.member_register("waweru@gmail.com", "mwaz", "password", "")
+        self.assertIn("205,Invalid input", result, "205,Invalid input")
+
+    def test_cpassword_is_equal_to_password(self):
+        """Test to check whether the confirm password and the passwords are similar """
+        result = self.newUser.member_register("waweru@gmail.com", "mwaz", "password", "pass")
+        self.assertIn("400,Passwords dont match", result, "400,Passwords dont match")
+
+    def test_cpassword_is_empty(self):
```
Comment (TinaMurimi) [inline test_users.py:51]: This test is not necessary and is a duplicate of `test_null_cpassword`. If the password doe not match the first one entered then that is handled by the test `test_cpassword_is_equal_to_password`

File: test_users.py [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+        """Test to check if member name is null"""
+        result = self.newUser.member_register("waweru@gmail.com", "", "password", "password")
+        self.assertIn("205,Regex mismatch", result, "205,Regex mismatch")
+
+    def test_null_password(self):
+        """Test to check if email is null"""
+        result = self.newUser.member_register("waweru@gmail.com", "mwaz", "", "password")
+        self.assertIn("205,Invalid input", result, "205,Invalid input")
+
+    def test_null_cpassword(self):
```
Comment (TinaMurimi) [inline test_users.py:41]: This test is not necessary. If the password doe not match the first one entered then that is handled by the test `test_cpassword_is_equal_to_password`

File: test_users.py [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+
+    def test_invalid_member_name(self):
+        """Test to check if member_name follows the one specified by regex"""
+        member_registration = self.newUser.member_register("waweru@gmail.com", "&#*", "password", "password")
+        self.assertIn("205,Regex mismatch", member_registration, "205,Regex mismatch")
+
+    def test_success_login(self):
+        """ method to test a successful login """
+        self.newUser.users = {}
+        self.newUser.member_register('waweru@gmail.com', 'mwaz', 'pass', 'pass')
```
Comment (TinaMurimi) [inline test_users.py:64]: This is used 3 times in your code. DRY! Should be in the `setUp()` method to avoid the repetition

File: test_users.py [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+        """ method to test if the email used for login is null"""
+        result = self.newUser.user_login('', 'pass')
+        self.assertIn("205,Invalid input", result, "205,Invalid input")
+
+    def test_null_login_password(self):
+        """ method to test if the password used for login is null """
+        result = self.newUser.user_login('waweru@gmail.com', '')
+        self.assertIn("205,Invalid input", result, "205,Invalid input")
+
+    def test_null_inputs(self):
```
Comment (TinaMurimi) [inline test_users.py:92]: This test is really not necessary

File: test_users.py [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+        """ method to test if both the password and login is null """
+        result = self.newUser.user_login('', '')
+        self.assertIn("205,Invalid input", result, "205,Invalid input")
+
+    def test_empty_email(self):
```
Comment (TinaMurimi) [inline test_users.py:97]: What is this test for?

File: test_users.py [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+        result = self.newUser.user_login(' ', 'pass')
+        self.assertIn("205,Empty input", result, "205,Empty input")
+
+    def test_empty_password(self):
```
Comment (TinaMurimi) [inline test_users.py:101]: What is this test for?

File: user.py [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+      """
+    def __init__(self, member=None, email=None, password=None):
+        """ Initializing  class instance variables"""
+        self.member = member
+        self.email = email
+        self.password = password
+
+    def member_register(self, email, member, password, cpassasword):
+        """ method that defines the elements required to create an account """
+        if re.match(regex_member, member):
```
Comment (TinaMurimi) [inline user.py:20]: Your `if` statements are untidy. If a condition is met it's body is executed, else it is ignored. So you can finish one `if` statement then continue with the next

File: user.py [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+                            return "205,Email Regex mismatch"
+                        return "400,Passwords dont match"
+                    return "401,Email exists"
+                return "401,Member exists"
+            return "205,Invalid input"
+        return "205,Regex mismatch"
+
+    def user_login(self, email, password):
+        """Method to login user"""
+        if email != ' ' and password != ' ':
```
Comment (TinaMurimi) [inline user.py:39]: `!= ' '`. This seems to occur everywhere in your code. Use `if not x: statement`

File: user.py [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+                    if passwd == password:
+                        return "200,OK"
+                    return "205,Password mismatch"
+                return "404,User not found"
+            return "205,Invalid input"
+        return "205,Empty input"
+
+
+    def get_member(self, email):
+        if email in users.keys():
```
Comment (TinaMurimi) [inline user.py:53]: Use ternary statements.

File: config.py [hunk 24]
```diff
@@ -0,0 +1,2 @@
+# Enable Flask's debugging features. Should be False in production
```
Comment (TinaMurimi) [inline config.py:1]: This should be in the `/app` folder

File: test_recipes.py [hunk 25]
```diff
@@ -0,0 +1,74 @@
+import unittest
+from recipe import Recipe
```
Comment (TinaMurimi) [inline test_recipes.py:2]: Why then do you have `/models/recipes.py`?  https://github.com/mwaz/yummy-recipies/tree/develop/models
